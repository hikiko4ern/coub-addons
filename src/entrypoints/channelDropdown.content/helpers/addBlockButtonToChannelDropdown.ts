import type { createAddChannelBlockButton } from '@/js/createAddChannelBlockButton';
import type { Logger } from '@/utils/logger';
import type { ChannelDropdownAddedNode, ChannelDropdownAddedNodes } from '../types';

const CHANNEL_FOLLOW_BUTTON_SELECTOR = 'div.channel-follow-button' as const;
const CHANNEL_ID_ATTR = 'data-channel-id' as const;
const CHANNEL_TITLE_SELECTOR = '.channel__title' as const;
const CHANNEL_LINK_SELECTOR = '.channel__title > a' as const;

export const addBlockButtonToChannelDropdown = (
	logger: Logger,
	addedNodes: ChannelDropdownAddedNodes,
	addChannelBlockButton: ReturnType<typeof createAddChannelBlockButton>,
	channelDropdownContent: Element,
	channelId?: number,
) => {
	logger.debug(
		'trying to get channel ID from attr',
		CHANNEL_ID_ATTR,
		'from node',
		channelDropdownContent,
	);

	if (typeof channelId !== 'number') {
		const channelIdAttr = channelDropdownContent.getAttribute(CHANNEL_ID_ATTR);
		const id = channelIdAttr && Number.parseInt(channelIdAttr, 10);

		if (!Number.isInteger(id)) {
			logger.warn('value', channelIdAttr, 'is not a valid channel ID');
			return;
		}

		channelId = id;
	}

	const nodeWithChannelTitle = channelDropdownContent.querySelector(CHANNEL_TITLE_SELECTOR);

	if (!nodeWithChannelTitle) {
		logger.info(
			'there is no node with channel title found by selector',
			CHANNEL_TITLE_SELECTOR,
			'in node',
			channelDropdownContent,
		);
		return;
	}

	const channelTitle =
		'innerText' in nodeWithChannelTitle && typeof nodeWithChannelTitle.innerText === 'string'
			? nodeWithChannelTitle.innerText
			: nodeWithChannelTitle.textContent;

	if (channelTitle === null) {
		logger.warn(
			'`innerText` of the channel title node',
			nodeWithChannelTitle,
			"isn't a string",
			channelTitle,
		);
		return;
	}

	let channelPermalink: string | undefined;

	try {
		const node = channelDropdownContent.querySelector(CHANNEL_LINK_SELECTOR);
		if (node) {
			channelPermalink = new URL(node.href).pathname.slice(1);
		} else {
			logger.warn(
				'there is no node with channel link found by selector',
				CHANNEL_LINK_SELECTOR,
				'in node',
				channelDropdownContent,
			);
		}
	} catch (err) {
		logger.error('failed to get channel permalink:', err);
	}

	const channelFollowButton = channelDropdownContent.querySelector(CHANNEL_FOLLOW_BUTTON_SELECTOR);

	if (!channelFollowButton) {
		logger.warn(
			'"Follow" button wrapper was not found with selector',
			CHANNEL_FOLLOW_BUTTON_SELECTOR,
			'in node',
			channelDropdownContent,
		);
		return;
	}

	const waivedWindow = window.wrappedJSObject || window;
	const { WeakRef } = waivedWindow;

	addChannelBlockButton({
		target: channelFollowButton,
		channel: {
			id: channelId,
			title: channelTitle,
			permalink: channelPermalink,
		},
		onAdded(blockButton, unsubscribe) {
			const entry = cloneInto([] as unknown as ChannelDropdownAddedNode, window);
			addedNodes.push(entry);

			entry[0] = new WeakRef(channelDropdownContent);
			entry[1] = new WeakRef(blockButton);
			entry[2] = new WeakRef(
				exportFunction(() => {
					logger.debug('removing channel', channelId, 'listener');
					unsubscribe();
				}, waivedWindow),
			);
		},
	});
};
