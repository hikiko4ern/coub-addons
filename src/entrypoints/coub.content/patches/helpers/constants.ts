import { COUB_PREFIX } from '../../constants';

export const APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY =
	`${COUB_PREFIX}__application__smartDateTime__orig` as const;
export const APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM = Symbol.for(
	APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY,
);
