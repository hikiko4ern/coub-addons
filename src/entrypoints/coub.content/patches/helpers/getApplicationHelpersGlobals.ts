import type { GetGlobalsRes } from '../types';

export interface ApplicationHelpersGlobals {
	moment: typeof window.moment;
	I18n: typeof window.I18n;
}

export const getApplicationHelpersGlobals = (): GetGlobalsRes<ApplicationHelpersGlobals> => {
	const { moment, I18n } = window.wrappedJSObject || window;

	if (typeof moment !== 'function') {
		return { isValid: false, ret: ['moment is not a function:', moment] };
	}

	if (typeof I18n !== 'object') {
		return { isValid: false, ret: ['`I18n` is not an object:', I18n] };
	}

	if (typeof I18n.t !== 'function') {
		return {
			isValid: false,
			ret: ['`I18n.t` is not a function:', I18n.t],
		};
	}

	return {
		isValid: true,
		ret: {
			moment,
			I18n,
		},
	};
};
