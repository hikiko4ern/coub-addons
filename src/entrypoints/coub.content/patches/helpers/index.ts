export { APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY } from './constants';
export { patchHelpers } from './patchHelpers';
export { revertHelpersPatches } from './revertHelpersPatches';
