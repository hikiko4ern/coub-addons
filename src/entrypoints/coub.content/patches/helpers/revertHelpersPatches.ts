import type {
	APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY,
	APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM,
} from './constants';

export function revertHelpersPatches(
	applicationOriginalSmartDateTimeKey: typeof APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY,
	loggerPrefix = '',
	logger: Pick<Console, 'debug' | 'error'> = console,
	helpers?: typeof window.helpers,
) {
	const log = (method: keyof typeof logger, ...args: unknown[]) =>
		loggerPrefix ? logger[method](`[${loggerPrefix}]`, ...args) : logger[method](...args);

	if (!helpers) {
		if (!('helpers' in window) || Reflect.getOwnPropertyDescriptor(window, 'helpers')?.set) {
			log('debug', 'removing helpers `defineProperty` patch');

			return Reflect.defineProperty(window, 'helpers', {
				configurable: true,
				enumerable: true,
				writable: true,
				value: undefined,
			});
		}

		helpers = window.helpers;
	}

	try {
		if (helpers && 'Application' in helpers) {
			const origSmartDateTimeSym: typeof APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM = Symbol.for(
				applicationOriginalSmartDateTimeKey,
			) as typeof APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM;

			const Application = helpers.Application;
			const origSmartDateTime = Application[origSmartDateTimeSym];

			if (origSmartDateTime) {
				Application.smartDateTime = origSmartDateTime;
				delete Application[origSmartDateTimeSym];
			}
		}
	} catch (err) {
		log('error', 'failed to revert `Application.smartDateTime` patch:', err);
	}
}
