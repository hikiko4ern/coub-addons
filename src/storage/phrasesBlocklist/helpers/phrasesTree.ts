import type { ToReadonly } from '@/types/util';
import type { SegmenterUtils } from './segmenterUtils';

export type PhrasesTreeLeaf = {
	[word in string]?: PhrasesTreeLeaf;
} & {
	[lastWord in LAST_WORD]?: string;
};

export interface PhrasesTree {
	[word: string]: PhrasesTreeLeaf;
}

// `structuredClone` does not preserve symbols, so we use a regular string with spaces as a key
// (segmentation separates words by spaces, so this key-marker should not overlap with key-words)
export const LAST_WORD = '#last word#';
export type LAST_WORD = typeof LAST_WORD;

export const phrasesToTree = (utils: SegmenterUtils, phrases: Iterable<string>): PhrasesTree => {
	const tree: PhrasesTree = {};

	for (let phrase of phrases) {
		phrase = preparePhraseForTree(phrase);
		const words = utils.segmentWords(phrase);

		if (words) {
			let leaf: PhrasesTreeLeaf = tree;

			for (const word of words) {
				leaf = leaf[word] ||= {};
			}

			leaf[LAST_WORD] = phrase;
		}
	}

	return tree;
};

export const preparePhraseForTree = (phrase: string) => phrase.normalize('NFKC').toLowerCase();

export const addPhraseToTree = (
	utils: SegmenterUtils,
	tree: ToReadonly<PhrasesTree>,
	phrase: string,
): ToReadonly<PhrasesTree> => {
	phrase = preparePhraseForTree(phrase);
	const words = utils.segmentWords(phrase);

	if (!words) {
		return tree;
	}

	const newTree: PhrasesTree = { ...tree };
	let leaf: PhrasesTreeLeaf = newTree;

	for (const word of words) {
		leaf = leaf[word] = word in leaf ? { ...leaf[word] } : {};
	}

	leaf[LAST_WORD] = phrase;

	return newTree;
};
