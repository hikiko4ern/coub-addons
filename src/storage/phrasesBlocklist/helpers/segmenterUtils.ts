import initWasm, { segmentWords as wasm_segmentWords } from '@coub-addons/segmenter-utils';
import createEmojiRegex from 'emoji-regex';

import type { MaybePromise } from '@/types/util';
import { Logger } from '@/utils/logger';

const logger = Logger.create('segmenter');
const wasmUrl = browser.runtime.getURL('segmenter-utils.wasm');

export interface SegmenterUtils {
	segmentWords: (input: string) => string[] | undefined;
}

let loadSegmenterUtils: () => MaybePromise<SegmenterUtils>;

if (typeof Intl.Segmenter === 'undefined') {
	logger.debug('using `Intl.Segmenter` polyfill');

	let initPromise: Promise<SegmenterUtils> | undefined;
	let segmenterUtils: SegmenterUtils | undefined;

	loadSegmenterUtils = () =>
		(initPromise ||= initWasm(wasmUrl).then((): SegmenterUtils => {
			logger.debug('successfully loaded from', wasmUrl);

			segmenterUtils = {
				segmentWords: wasm_segmentWords,
			};

			// biome-ignore lint/style/noNonNullAssertion: it's guaranteed to be already loaded
			loadSegmenterUtils = () => segmenterUtils!;

			return segmenterUtils;
		})).catch((err: unknown) => {
			logger.error('failed to load from', wasmUrl, err);
			throw err;
		});
} else {
	logger.debug('using native `Intl.Segmenter`');

	const segmenter = new Intl.Segmenter(undefined, { granularity: 'word' });
	const emojiRegex = createEmojiRegex();

	const segmenterUtils: SegmenterUtils = {
		segmentWords: function segmentWords(input: string) {
			const words: string[] = [];

			for (const s of segmenter.segment(input)) {
				if (s.isWordLike || emojiRegex.test(s.segment)) {
					words.push(s.segment);
				}
			}

			return words.length ? words : undefined;
		},
	};

	loadSegmenterUtils = () => segmenterUtils;
}

export { loadSegmenterUtils };
