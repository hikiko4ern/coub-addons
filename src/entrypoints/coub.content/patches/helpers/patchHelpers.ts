import { isObject } from '@/helpers/isObject';
import { type RevertPatch, applyPatches } from '@/helpers/patch/applyPatches';
import type { Logger } from '@/utils/logger';

import {
	APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY,
	APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM,
} from './constants';
import { getApplicationHelpersGlobals } from './getApplicationHelpersGlobals';
import { revertHelpersPatches } from './revertHelpersPatches';

type xApplicationPatches = {
	[key in typeof APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM]?: coub.helpers.Application['smartDateTime'];
};

declare global {
	namespace coub {
		namespace helpers {
			interface ApplicationPatches extends xApplicationPatches {}
		}
	}
}

export function patchHelpers(
	parentLogger: Logger,
	waivedWindow: typeof window,
): RevertPatch | unknown[] {
	const logger = parentLogger.getChildLogger('helpers');

	const helpers = waivedWindow.helpers;

	const patches = applyPatches(
		logger,
		helpers,
		{
			Application: patchApplication,
		},
		waivedWindow,
	);

	logger.debug('patched successfully');

	return () => {
		logger.debug('removing patches');

		for (const revert of patches) {
			revert?.();
		}

		revertHelpersPatches(
			APPLICATION_ORIGINAL_SMART_DATE_TIME_KEY,
			undefined,
			logger,
			waivedWindow.helpers,
		);
	};
}

const patchApplication = (parentLogger: Logger, waivedWindow: typeof window) => {
	const validatedGlobals = getApplicationHelpersGlobals();

	if (!validatedGlobals.isValid) {
		return validatedGlobals.ret;
	}

	const logger = parentLogger.getChildLogger('Application');
	const { moment, I18n } = validatedGlobals.ret;
	const Application = waivedWindow.helpers.Application;

	{
		const origSmartDateTime = Application[APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM];

		if (typeof origSmartDateTime === 'function') {
			logger.debug('reverting non-reverted `smartDateTime` patch');
			Application.smartDateTime = origSmartDateTime;
			delete Application[APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM];
		}
	}

	const origSmartDateTime = (Application[APPLICATION_ORIGINAL_SMART_DATE_TIME_SYM] =
		Application.smartDateTime);

	type smartDateTime = (typeof Application)['smartDateTime'];

	const smartDateTimeLogger = logger.getChildLogger('smartDateTime');

	const patchedSmartDateTime: smartDateTime = function patchedSmartDateTime(
		this: ThisParameterType<smartDateTime>,
		...args
	) {
		smartDateTimeLogger.debug('called with', [...args]);

		const [object, _ago, dateProperty = 'created_at'] = args;

		if (
			isObject(object) &&
			typeof dateProperty === 'string' &&
			dateProperty in object &&
			typeof object[dateProperty as keyof typeof object] === 'string'
		) {
			try {
				const rawDate = object[dateProperty as keyof typeof object] as string,
					now = new Date(),
					date = moment(rawDate),
					dateYear = date.year();

				logger.debug('created date');

				if (dateYear !== now.getFullYear()) {
					return I18n.t(
						'smart_datetime_new.year_ago',
						cloneInto(
							{
								day: date.date(),
								month: I18n.t('smart_datetime_new.month_names')[date.month()],
								year: dateYear,
							},
							I18n,
						),
					);
				}
			} catch (err) {
				smartDateTimeLogger.error(
					'failed to format date, falling back to the provided implementation',
					err,
				);
			}
		}

		return Reflect.apply(origSmartDateTime, this, args);
	};

	exportFunction(patchedSmartDateTime, Application, { defineAs: 'smartDateTime' });

	logger.debug('patched successfully');
};
