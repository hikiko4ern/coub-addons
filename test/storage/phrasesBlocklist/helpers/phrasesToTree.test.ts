// cspell:words hewwo, wowwd, beautifuw

import { permutations } from 'itertools';
import { expect, it } from 'vitest';

import { LAST_WORD, phrasesToTree } from '@/storage/phrasesBlocklist/helpers/phrasesTree';
import { segmenterUtils } from './segmenterUtils';

it('1 word', () => {
	expect(phrasesToTree(segmenterUtils, ['hewwo'])).toStrictEqual({
		hewwo: { [LAST_WORD]: 'hewwo' },
	});

	expect(phrasesToTree(segmenterUtils, ['hewwo!'])).toStrictEqual({
		hewwo: { [LAST_WORD]: 'hewwo!' },
	});

	expect(phrasesToTree(segmenterUtils, ['hewwo', 'hewwo'])).toStrictEqual({
		hewwo: { [LAST_WORD]: 'hewwo' },
	});
});

it('2 words', () => {
	expect(phrasesToTree(segmenterUtils, ['hewwo wowwd'])).toStrictEqual({
		hewwo: {
			wowwd: { [LAST_WORD]: 'hewwo wowwd' },
		},
	});
});

it('3 words', () => {
	expect(phrasesToTree(segmenterUtils, ['hewwo beautifuw wowwd'])).toStrictEqual({
		hewwo: {
			beautifuw: {
				wowwd: { [LAST_WORD]: 'hewwo beautifuw wowwd' },
			},
		},
	});
});

it('1 + 2 words', () => {
	for (const words of permutations(['hewwo', 'hewwo wowwd'])) {
		expect(phrasesToTree(segmenterUtils, words)).toStrictEqual({
			hewwo: {
				[LAST_WORD]: 'hewwo',
				wowwd: { [LAST_WORD]: 'hewwo wowwd' },
			},
		});
	}
});

it('1 + 2 + 3 words', () => {
	for (const words of permutations(['hewwo', 'hewwo wowwd', 'hewwo beautifuw wowwd'])) {
		expect(phrasesToTree(segmenterUtils, words)).toStrictEqual({
			hewwo: {
				[LAST_WORD]: 'hewwo',
				wowwd: { [LAST_WORD]: 'hewwo wowwd' },
				beautifuw: {
					wowwd: { [LAST_WORD]: 'hewwo beautifuw wowwd' },
				},
			},
		});
	}
});

it('sentence', () => {
	const sentence =
		'Кроличья нора сначала была прямая, как тоннель, но потом обрывалась так внезапно';

	expect(phrasesToTree(segmenterUtils, [sentence])).toStrictEqual({
		кроличья: {
			нора: {
				сначала: {
					была: {
						прямая: {
							как: {
								тоннель: {
									но: {
										потом: {
											обрывалась: {
												так: {
													внезапно: { [LAST_WORD]: sentence.toLocaleLowerCase('ru') },
												},
											},
										},
									},
								},
							},
						},
					},
				},
			},
		},
	});
});

it('many words', () => {
	expect(
		phrasesToTree(segmenterUtils, [
			// I will definitely burn in hell for this
			// cspell:disable-next-line
			'Teh rabbit-hwowal went stwaight on wike a tunnyew fwow swome way (o´∀`o) and when dipped suddenwy dwown (* ^ ω ^) swo suddenwy dat Awice had nyot a mwoment two fwink abwout stwopping hewsewf befwowe she fwound hewsewf fawwing dwown a wewwy deep weww UwU',
		]),
	).toMatchSnapshot();
});
