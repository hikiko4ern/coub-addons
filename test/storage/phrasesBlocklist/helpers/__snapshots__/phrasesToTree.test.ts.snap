// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`many words 1`] = `
{
  "teh": {
    "rabbit": {
      "hwowal": {
        "went": {
          "stwaight": {
            "on": {
              "wike": {
                "a": {
                  "tunnyew": {
                    "fwow": {
                      "swome": {
                        "way": {
                          "o": {
                            " ́": {
                              "o": {
                                "and": {
                                  "when": {
                                    "dipped": {
                                      "suddenwy": {
                                        "dwown": {
                                          "*": {
                                            "ω": {
                                              "swo": {
                                                "suddenwy": {
                                                  "dat": {
                                                    "awice": {
                                                      "had": {
                                                        "nyot": {
                                                          "a": {
                                                            "mwoment": {
                                                              "two": {
                                                                "fwink": {
                                                                  "abwout": {
                                                                    "stwopping": {
                                                                      "hewsewf": {
                                                                        "befwowe": {
                                                                          "she": {
                                                                            "fwound": {
                                                                              "hewsewf": {
                                                                                "fawwing": {
                                                                                  "dwown": {
                                                                                    "a": {
                                                                                      "wewwy": {
                                                                                        "deep": {
                                                                                          "weww": {
                                                                                            "uwu": {
                                                                                              "#last word#": "teh rabbit-hwowal went stwaight on wike a tunnyew fwow swome way (o ́∀\`o) and when dipped suddenwy dwown (* ^ ω ^) swo suddenwy dat awice had nyot a mwoment two fwink abwout stwopping hewsewf befwowe she fwound hewsewf fawwing dwown a wewwy deep weww uwu",
                                                                                            },
                                                                                          },
                                                                                        },
                                                                                      },
                                                                                    },
                                                                                  },
                                                                                },
                                                                              },
                                                                            },
                                                                          },
                                                                        },
                                                                      },
                                                                    },
                                                                  },
                                                                },
                                                              },
                                                            },
                                                          },
                                                        },
                                                      },
                                                    },
                                                  },
                                                },
                                              },
                                            },
                                          },
                                        },
                                      },
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
}
`;
